### PR TITLE
Convert logger to typescript

### PR DESCRIPTION
**Pull request by marcusramberg**

### CODE REVIEW

File: src/logger.ts [hunk 1]
```diff
@@ -1,7 +1,12 @@
+import type {WriteStream} from 'fs';
 import assert from 'assert';
 import chalk from 'chalk';
 
-const LEVEL = Object.freeze({
+export type LogFormatter = (data: LogEvent) => string;
+export interface LogContext { requestId?: string }
+export interface LogEvent { level: string, time: string, requestId?: string, msg: string }
```
Comment (kraih) [inline src/logger.ts:7]: Why are they exported?

File: src/logger.ts [hunk 2]
```diff
@@ -1,7 +1,12 @@
+import type {WriteStream} from 'fs';
 import assert from 'assert';
 import chalk from 'chalk';
 
-const LEVEL = Object.freeze({
+export type LogFormatter = (data: LogEvent) => string;
+export interface LogContext { requestId?: string }
```
Comment (kraih) [inline src/logger.ts:6]: The context is an arbitraty object, `requestId` is just the one our default formatter uses. It's a Node logging standard.

File: src/logger.ts [hunk 3]
```diff
@@ -1,7 +1,12 @@
...
+export interface LogEvent { level: string, time: string, requestId?: string, msg: string }
+
+const LEVEL: Record<string, number> = Object.freeze({
```
Comment (kraih) [inline src/logger.ts:9]: Is the type really not inferred here? How odd.

File: src/logger.ts [hunk 4]
```diff
@@ -11,75 +16,87 @@ const LEVEL = Object.freeze({
...
 export default class Logger {
-  constructor (options = {}) {
-    this.formatter = options.formatter || Logger.colorFormatter;
-    this.destination = options.destination || process.stderr;
+  formatter: LogFormatter;
+  destination: NodeJS.WritableStream;
+  private _level: number;
+  history: LogEvent[];
+  private readonly _historySize: number;
+  constructor (options: {
```
Comment (kraih) [inline src/logger.ts:24]: Missing blank line before constructor. And we don't really mark underscore properties as `private` or `readonly`.

File: src/logger.ts (hunk 2 above)
Comment (kraih) [reply src/logger.ts:6]: You can keep `requestId?`, but it also needs to allow arbitrary pairs.

File: src/logger.ts [hunk 5]
```diff
@@ -11,75 +16,87 @@ const LEVEL = Object.freeze({
...
-  static jsonFormatter (data) {
+  static jsonFormatter (data: LogEvent): string {
     return JSON.stringify(data);
   }
 
-  get level () {
-    return Object.keys(LEVEL).find(key => LEVEL[key] === this._level);
+  get level (): string {
+    const level = Object.keys(LEVEL).find(key => LEVEL[key] === this._level);
+    assert(level !== undefined, `Unsupported log level "${this._level}"`);
```
Comment (kraih) [inline src/logger.ts:81]: This assertion needs a test case.

File: src/logger.ts [hunk 6]
```diff
@@ -97,32 +114,34 @@ export default class Logger {
 }
 
 export class ChildLogger {
-  constructor (parent, context) {
+  parent: Logger;
+  context: LogContext;
+  constructor (parent: Logger, context: LogContext) {
```
Comment (kraih) [inline src/logger.ts:119]: Missing blank line again.

File: src/plugins/exception-helpers.ts [hunk 7]
```diff
@@ -42,7 +42,7 @@ async function htmlNotFound (ctx: MojoContext): Promise<boolean> {
 }
 
 async function httpException (ctx: MojoContext, error: Error): Promise<boolean> {
-  ctx.log.error(error.stack);
+  ctx.log.error(error.stack ?? 'error.stack is undefined for ' + error.message);
```
Comment (kraih) [inline src/plugins/exception-helpers.ts:45]: Why does the user care if the stack is missing?

Comment (kraih) [reply src/plugins/exception-helpers.ts:45]: Also would need a test case.

File: src/logger.ts (hunk 1 above)
Comment (marcusramberg) [reply src/logger.ts:7]: I was wrongly thinking they needed to be to be used externally, but I guess only the function signatures will be sufficient.

File: src/logger.ts (hunk 3 above)
Comment (marcusramberg) [reply src/logger.ts:9]: This happens when it's not defined:
image

File: src/plugins/exception-helpers.ts (hunk 7 above)
Comment (marcusramberg) [reply src/plugins/exception-helpers.ts:45]: changed to `as string`

File: src/logger.ts (hunk 5 above)
Comment (marcusramberg) [reply src/logger.ts]: changed to `as string` since setter already has the assert.

**Pull request merged**